### PR TITLE
JAMES-2346 JMAP should be able to send MDNs

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        MailboxProbe mailboxProbe = jmapServer.getProbe(MailboxProbeImpl.class);
+        DataProbe dataProbe = jmapServer.getProbe(DataProbeImpl.class);
+
+        RestAssured.requestSpecification = new RequestSpecBuilder()
+                .setContentType(ContentType.JSON)
+                .setAccept(ContentType.JSON)
+                .setConfig(newConfig().encoderConfig(encoderConfig().defaultContentCharset(StandardCharsets.UTF_8)))
+                .setPort(jmapServer.getProbe(JmapGuiceProbe.class).getJmapPort())
+                .build();
+        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java]: I'm not really found of this, because when you are sending jmap requests in await method in implementations (in order to know if James is up, for example) then a lot of logs are appended.

then you can't load the full logs from CI results...

File: mdn/src/test/java/org/apache/james/mdn/MDNTest.java [hunk 2]
```diff
@@ -164,4 +166,36 @@ public void humanReadableTextShouldNotBeTrimmed() throws Exception {
...
+                "Explanation:\n" +
+                " - We should always write detailed unit tests\n" +
+                " - We should think of all edge cases")
+        .contains("Content-Type: message/disposition-notification; charset=UTF-8\r\n" +
+            "\r\n" +
+            "Final-Recipient: rfc822; final@domain.com\r\n" +
+            "Disposition: automatic-action/MDN-sent-automatically;deleted");
+    }
+
+    public String asString(Message message) throws Exception {
```
Comment (aduprat) [inline mdn/src/test/java/org/apache/james/mdn/MDNTest.java]: private

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java [hunk 3]
```diff
@@ -266,4 +276,31 @@ public void gpgSignatureShouldBeConsideredAsAnAttachment() throws Exception {
             .extracting(Optional::get)
             .containsOnly("message suivi", "signature.asc");
     }
+
+    @Test
+    public void mdnReportsShouldBeConsideredAs() throws Exception {
```
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: attachmentsShouldBeConsideredAsMDNWhenDispositionContentType

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java [hunk 4]
```diff
@@ -93,11 +99,37 @@ public void setup() throws Throwable {
         dataProbe.addUser(BOB, BOB_PASSWORD);
         mailboxProbe.createMailbox("#private", USERNAME, DefaultMailboxes.INBOX);
         accessToken = HttpJmapAuthentication.authenticateJamesUser(baseUri(), USERNAME, PASSWORD);
+        bobAccessToken = HttpJmapAuthentication.authenticateJamesUser(baseUri(), BOB, BOB_PASSWORD);
+        await();
+
+        String messageCreationId = "creationId";
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java]: Can you extract this provisioning to a method & call it on each test methods if needed.

Comment (aduprat) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java]: this method can return the created MessageId.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java [hunk 5]
```diff
@@ -21,38 +21,195 @@
...
-                    .build()));
+        messageSender.sendMessage(metaDataWithContent,
+            Envelope.fromMime4JMessage(mdnAnswer), mailboxSession);
 
-        return builder.build();
+        return metaDataWithContent.getMessageId();
     }
+
+    public Message generateMDNMessage(Message originalMessage, JmapMDN mdn, MDNReport mdnReport, User user) throws ParseException, IOException {
+        return MDN.builder()
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: Can you change the indent, something like:
```
MDN.builder()
            .report(mdnReport)
            .humanReadableText(mdn.getTextBody())
            .build()
    .asMime4JMessageBuilder()
            .setTo(originalMessage.getSender().getAddress())
            .setFrom(user.asString())
            .setSubject(mdn.getSubject())
            .setMessageId(MimeUtil.createUniqueMessageId(user.getDomainPart().orElse(null)))
            .build()
```

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java [hunk 6]
```diff
@@ -21,38 +21,195 @@
...
-                    .type("Not implemented yet")
-                    .build()));
+        messageSender.sendMessage(metaDataWithContent,
+            Envelope.fromMime4JMessage(mdnAnswer), mailboxSession);
 
-        return builder.build();
+        return metaDataWithContent.getMessageId();
     }
+
+    public Message generateMDNMessage(Message originalMessage, JmapMDN mdn, MDNReport mdnReport, User user) throws ParseException, IOException {
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: Maybe private, no? (same for other methods)

File: server/protocols/jmap/pom.xml [hunk 7]
```diff
@@ -80,6 +80,10 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>apache-mime4j-dom</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>james-mdn</artifactId>
```
Comment (mbaechler) [inline server/protocols/jmap/pom.xml]: mdn is a bad project name : when looking to the web for mdn, you find everything but the right RFC.
Please use either the complete name "Message Disposition Notification" or the functional description "Message Receipts"

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/ValueWithId.java [hunk 8]
```diff
@@ -50,6 +51,12 @@ public CreationMessageEntry(CreationMessageId creationId, CreationMessage messag
         }
     }
 
+    public static class CreationMDNEntry extends ValueWithId<MDN> {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/ValueWithId.java]: MDNCreationEntry

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java [hunk 9]
```diff
@@ -21,38 +21,195 @@
...
 
 public class SendMDNProcessor implements SetMessagesProcessor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SendMDNProcessor.class);
+    private static final MimeConfig MIME_ENTITY_CONFIG = MimeConfig.custom()
+        .setMaxContentLen(-1)
+        .setMaxHeaderCount(-1)
+        .setMaxHeaderLen(-1)
+        .setMaxLineLen(-1)
+        .build();
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: time to put this constant somewhere

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java [hunk 10]
```diff
@@ -21,38 +21,195 @@
...
     }
+
+    public Message generateMDNMessage(Message originalMessage, JmapMDN mdn, MDNReport mdnReport, User user) throws ParseException, IOException {
+        return MDN.builder()
+            .report(mdnReport)
+            .humanReadableText(mdn.getTextBody())
+            .build()
+            .asMime4JMessageBuilder()
+            .setTo(originalMessage.getSender().getAddress())
+            .setFrom(user.asString())
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: user? I think we don't support no virtual hosting in james, but it seems to be a strange shortcut to me.

Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: should be recipient?

**Comment by aduprat** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project cassandra-jmap-integration-testing: There was a timeout or other error in the fork -> [Help 1]
```
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java [hunk 11]
```diff
@@ -212,6 +226,11 @@ public String getNewState() {
         return notDestroyed;
     }
 
+    @JsonProperty("MDNSent")
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java]: needed?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 12]
```diff
@@ -57,7 +57,6 @@
         .setMaxContentLen(-1)
         .setMaxHeaderCount(-1)
         .setMaxHeaderLen(-1)
-        .setMaxHeaderCount(-1)
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: This line was duplicated

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java (hunk 11 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java]: Yes, to match the case of the spec

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java [hunk 13]
```diff
@@ -196,13 +212,19 @@ public String getNewState() {
         return notDestroyed;
     }
 
+    @JsonProperty("MDNNotSent")
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java]: Yes, to match the case of the spec (Jackson would return "mdnNotSent" otherwise)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: More mdnReportShouldBeConsideredAsAttachmentWhenDispositionContentType

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java (hunk 6 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: Looks like a bad IDE setting...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/SetMessagesResponse.java]: Answered already

**Comment by chibenwa** [comment]
Comments that needs to be taken into account:

 - https://github.com/linagora/james-project/pull/1311#discussion_r173194075 Use of **User** as mail address

I'm not aware of an alternative.... In normal message creation, this is provided by the MUA.

 - https://github.com/linagora/james-project/pull/1311#discussion_r173190393 MIME4J constant to extract. Good place is in upstream project, but this is a bit "heavy".

 - https://github.com/linagora/james-project/pull/1311#discussion_r173182130 Missing test for Mime4J message -> bytes convertion. Might also deserve to be in a helper method in Mime4J project...

 - https://github.com/linagora/james-project/pull/1311#discussion_r173180927 Abusive use of await. Should be polling instead

 - https://github.com/linagora/james-project/pull/1311#discussion_r173179379 Fails with NPE if no sender

 - https://github.com/linagora/james-project/pull/1311#discussion_r173171070 **james-mdn** naming not considered "enough"...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java [hunk 14]
```diff
@@ -21,38 +21,195 @@
...
+        return metaDataWithContent.getMessageId();
     }
+
+    public Message generateMDNMessage(Message originalMessage, JmapMDN mdn, MDNReport mdnReport, User user) throws ParseException, IOException {
+        return MDN.builder()
+            .report(mdnReport)
+            .humanReadableText(mdn.getTextBody())
+            .build()
+            .asMime4JMessageBuilder()
+            .setTo(originalMessage.getSender().getAddress())
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: Done, and I extracted it for easy unit testing.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java (hunk 10 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: Ok, having a look at **SetMessagesCreationProcessor::assertUserIsSender (l 270)** we already assume that the username is the user mail address:

```
    private void assertUserIsSender(MailboxSession session, Optional<DraftEmailer> from) throws MailboxSendingNotAllowedException {
        if (!from.flatMap(DraftEmailer::getEmail)
                .filter(email -> session.getUser().isSameUser(email))
                .isPresent()) {
            String allowedSender = session.getUser().getUserName();
            throw new MailboxSendingNotAllowedException(allowedSender);
        }
    }
```

I would thus say we don't need to worry about this.

**Comment by chibenwa** [comment]
I will create the PR for mime4J requested changes.

I propose we uncouple this PR from Mime4J codebase changes... I will handle that in a separated PR.

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-mime4j/pull/12 . That will lead to an Upcoming PR to integrate it in James source code.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: See https://github.com/linagora/james-mime4j/pull/12

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MIMEMessageConverter.java [hunk 15]
```diff
@@ -107,11 +107,14 @@ public MIMEMessageConverter() {
     }
 
     public byte[] convert(ValueWithId.CreationMessageEntry creationMessageEntry, ImmutableList<MessageAttachment> messageAttachments) {
+        return asBytes(convertToMime(creationMessageEntry, messageAttachments));
+    }
 
+    public byte[] asBytes(Message message) {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MIMEMessageConverter.java]: See https://github.com/linagora/james-mime4j/pull/12

File: server/protocols/jmap/pom.xml (hunk 7 above)
Comment (chibenwa) [reply server/protocols/jmap/pom.xml]: Sorry but I'm :-1: ...

"Receipts" means nothing to me. This term doesn't even appear in the codebase. And "receipt" have only 3 match on the RFC page. Seems like a usage of MDNs but not a good qualification of it.

Honestly, **james-message-disposition-notification** seems ridiculous to me... It's as if we had **james-sender-policy-framework**, **james-server-json-metadata-application-protocol**, **james-server-simple-messaging-transfer-protocol**, etc... It don't seems to be a project convention to me...

If I search **RFC MDN** my first page is https://tools.ietf.org/html/rfc3798 confirming that **mdn** is a better fit.

Furthermore, this naming had already been validated in a previous review...

Also with such a change, we might brake external potential usage...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java [hunk 16]
```diff
@@ -141,7 +210,131 @@ public void sendMDNIsNotSupportedYet() {
...
+                "    \"disposition\":{" +
+                "        \"actionMode\":\"" + DispositionActionMode.Automatic.getValue() + "\","+
+                "        \"sendingMode\":\"" + DispositionSendingMode.Automatic.getValue() + "\","+
+                "        \"type\":\"" + DispositionType.Processed.getValue() + "\""+
+                "    }" +
+                "}" +
+                "}}, \"#0\"]]")
+            .post("/jmap");
+
+        await();
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java]: Strictly speaking it's not. I fixed it.

File: server/protocols/jmap/pom.xml (hunk 7 above)
Comment (mbaechler) [reply server/protocols/jmap/pom.xml]: > Furthermore, this naming had already been validated in a previous review...

Not a valid argument, it's very usual to miss things during review

> If I search RFC MDN my first page is https://tools.ietf.org/html/rfc3798 confirming that mdn is a better fit.

Ok but nobody prefixes searchs by RFC without any advice to do so

> Honestly, james-message-disposition-notification seems ridiculous to me... It's as if we had james-sender-policy-framework, james-server-json-metadata-application-protocol, james-server-simple-messaging-transfer-protocol, etc... It don't seems to be a project convention to me...

Your examples are interesting : the first search results for SPF, JSON and SMTP gives me a link to sensible information. MDN doesn't. If you try to search "mdn mail" because James is about mails, you don't find the right information neither. So your argument doesn't hold IMO.

Comment (chibenwa) [reply server/protocols/jmap/pom.xml]: > Ok but nobody prefixes searchs by RFC without any advice to do so

Me I do... It refine the search quitte well ....

> Your examples are interesting : the first search results for SPF, JSON and SMTP gives me a link to sensible information. MDN doesn't. If you try to search "mdn mail" because James is about mails, you don't find the right information neither. So your argument doesn't hold IMO.

I don't see why we should do something different than in the general case here...

Maybe we need RFC number in the **pom** description, but I don't think a complicated and long artifact ID is the solution to a lack of knowledge.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java (hunk 10 above)
Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SendMDNProcessor.java]: The coupling is hard. Can we add a test to ensure this does not break when we remove this constraint? (for example when we'll implement identities)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.exceptions;
+
+public class InvalidOriginMessageForMDNNotificationException extends Exception {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java]: s/MDNNotification/MDN/ ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java [hunk 18]
```diff
@@ -116,6 +131,55 @@ public MDNDisposition getDisposition() {
...
+            .setMessageId(MimeUtil.createUniqueMessageId(user.getDomainPart().orElse(null)))
+            .build();
+    }
+
+    public String getSenderAddress(Message originalMessage) throws InvalidOriginMessageForMDNNotificationException {
+        return OptionalUtils.or(
+            Optional.ofNullable(originalMessage.getSender()),
+            Optional.ofNullable(originalMessage.getFrom())
+                .map(MailboxList::stream).orElse(Stream.of())
+                .findAny())
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: in MessageFactory we findFirst()

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/JmapMDNTest.java [hunk 19]
```diff
@@ -122,4 +128,50 @@ public void textBodyIsCompulsory() {
...
+            .setBody("body", StandardCharsets.UTF_8)
+            .build();
+
+        assertThatThrownBy(() ->
+            MDN.generateMDNMessage(originMessage,
+                new MockMailboxSession("user@localhost.com"))
+                .getTo())
+            .isInstanceOf(InvalidOriginMessageForMDNNotificationException.class);
+    }
+
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/JmapMDNTest.java]: you miss some tests: sender + nofrom, sender + from, multiple sender (?), multiple from

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java (hunk 17 above)
Comment (fvn-linagora) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java]: Missing(Sender|Origin) i/of Invalid ?

**Comment by chibenwa** [comment]
```

[c046ffd05f1183b527010fa94cb09f3d1b4f2e16] Failed tests: 
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16]   ObjectMapperFactoryTest.readValueShouldFailOnInvalidDispositionType:204 
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16] Expecting message:
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16]  <"Unrecognized MDN Disposition sending mode illegal. Should be one of [deleted, dispatched, displayed, processed]">
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16] but was:
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16]  <"Unrecognized MDN Disposition type illegal. Should be one of [deleted, dispatched, displayed, processed]">
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16] 
[c046ffd05f1183b527010fa94cb09f3d1b4f2e16] Tests run: 900, Failures: 1, Errors: 0, Skipped: 2
```

Of course...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java (hunk 18 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: If you want, both seems correct to me. (When do we have several **From** entries?)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java (hunk 17 above)
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java]: @fvn-linagora  the idea here is to be flexible, and assume the exception class name should say that the origin message is invalid

This exception is designed to be more flexible than just "X is missing".

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/JmapMDNTest.java (hunk 19 above)
Comment (chibenwa) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/model/JmapMDNTest.java]: No API for multi senders, I won't test that case here, if you don't mind...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java (hunk 17 above)
Comment (fvn-linagora) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/InvalidOriginMessageForMDNNotificationException.java]: I'd say *YAGNI* ;) Better be explicit with guard clauses, in general. Only found usages through factory method `missingField`

**Comment by chibenwa** [comment]
I guess that would not be silly to do Reply-To then Sender then From....

To do that we can introduce a `Optional<T> OptionalUtils::or(Optional<T>... optionals)`

**Comment by chibenwa** [comment]
```
[ac6bb5f20db51ed086a9c324cd469c725328c91a] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 12.13 sec - in org.apache.james.mpt.managesieve.cassandra.CassandraCheckScriptTest
[ac6bb5f20db51ed086a9c324cd469c725328c91a] 
[ac6bb5f20db51ed086a9c324cd469c725328c91a] Results :
[ac6bb5f20db51ed086a9c324cd469c725328c91a] 
[ac6bb5f20db51ed086a9c324cd469c725328c91a] Tests in error: 
[ac6bb5f20db51ed086a9c324cd469c725328c91a]   HaveSpaceTest.haveSpaceShouldWork » IllegalState Can't use this cluster instan...
```

Seems unrelated...

**Comment by chibenwa** [comment]
Here are "anonymized" headers of one of my mails:

```
Return-Path: <someone@linagora.com>
Message-ID: <5469B169.3000701@linagora.com>
Date: Mon, 17 Nov 2014 09:27:21 +0100
From: Someone <someone@linagora.com>
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:31.0) Gecko/20100101 Thunderbird/31.1.2
MIME-Version: 1.0
To: Benoit Tellier <receiver@linagora.com>
Subject: Re: Contact hangouts
References: <JWC-james-mailbox-1@Stash> <JWC-james-mailbox-1-1313460438043688@Stash> <5469ADD2.6030306@linagora.com>
In-Reply-To: <5469ADD2.6030306@linagora.com>
Content-Type: text/plain; charset=utf-8; format=flowed
Content-Transfer-Encoding: 8bit
```

We see we can rely on the `ReturnPath` header.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java [hunk 20]
```diff
@@ -101,7 +97,9 @@ public void setup() throws Throwable {
         accessToken = HttpJmapAuthentication.authenticateJamesUser(baseUri(), USERNAME, PASSWORD);
         bobAccessToken = HttpJmapAuthentication.authenticateJamesUser(baseUri(), BOB, BOB_PASSWORD);
         await();
+    }
 
+    public void sendAnInitialMessage() {
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SendMDNMethodTest.java]: _private_ also on _getMessageIdListForAccount_ & _listMessagesInMailbox_

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java [hunk 21]
```diff
@@ -116,6 +131,55 @@ public MDNDisposition getDisposition() {
...
+            .build()
+        .asMime4JMessageBuilder()
+            .setTo(getSenderAddress(originalMessage))
+            .setFrom(user.asString())
+            .setSubject(subject)
+            .setMessageId(MimeUtil.createUniqueMessageId(user.getDomainPart().orElse(null)))
+            .build();
+    }
+
+    public String getSenderAddress(Message originalMessage) throws InvalidOriginMessageForMDNNotificationException {
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: private

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java [hunk 22]
```diff
@@ -116,6 +131,55 @@ public MDNDisposition getDisposition() {
...
+        return OptionalUtils.or(
+            Optional.ofNullable(originalMessage.getSender()),
+            Optional.ofNullable(originalMessage.getFrom())
+                .map(MailboxList::stream).orElse(Stream.of())
+                .findAny())
+            .orElseThrow(() -> InvalidOriginMessageForMDNNotificationException.missingField("Sender"))
+            .getAddress();
+    }
+
+    @JsonIgnore
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: Why do you need this annotation here and not on the _generateMDNMessage_ method?

Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: Probably not needed at all.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java [hunk 23]
```diff
@@ -148,15 +149,22 @@ public Message generateMDNMessage(Message originalMessage, MailboxSession mailbo
     }
 
     public String getSenderAddress(Message originalMessage) throws InvalidOriginMessageForMDNException {
-        return OptionalUtils.or(
-            Optional.ofNullable(originalMessage.getSender()),
-            Optional.ofNullable(originalMessage.getFrom())
-                .map(MailboxList::stream).orElse(Stream.of())
-                .findFirst())
+        Optional<Mailbox> replyTo = Optional.ofNullable(originalMessage.getReplyTo())
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapMDN.java]: Good catch.....

I need to fix this then....

**Comment by chibenwa** [comment]
```
[69e6296104660a7fc2ecbe0b4976bd6bf44b0106] Error executing docker rm --force=true --volumes=true git-publish-james-2346-00f28eb3-e325-4d1c-94dc-6da741f064f9 : no known node for /home/jenkins/build/workspace/95
```

Strange... test this please

**Comment by rouazana** [comment]
test this please

**Comment by chibenwa** [comment]
Merged
